src/web/crate_details.rs

```diff
@@ -293,75 +299,87 @@ struct CrateDetailsPage {
+    Extension(pool): Extension<Pool>,
+    Extension(repository_stats_updater): Extension<Arc<RepositoryStatsUpdater>>,
+) -> AxumResult<AxumResponse> {
+    // this handler must always called with a crate name
+    if params.version.is_none() {
+        return Ok(super::axum_cached_redirect(
+            &format!("/crate/{}/latest", params.name),
+            CachePolicy::ForeverInCdn,
+        )
+        .into_response());
     }
 
-    let mut conn = extension!(req, Pool).get()?;
+    let found_version = spawn_blocking({
+        let pool = pool.clone();
+        let params = params.clone();
+        move || {
+            let mut conn = pool.get()?;
+            match_version(&mut conn, &params.name, params.version.as_deref())
+                .and_then(|m| m.assume_exact())
+                .context("error matching version")
+        }
+    })
+    .await?;
 
-    let found_version =
-        match_version(&mut conn, name, req_version).and_then(|m| m.assume_exact())?;
     let (version, version_or_latest, is_latest_url) = match found_version {
         MatchSemver::Exact((version, _)) => (version.clone(), version, false),
         MatchSemver::Latest((version, _)) => (version, "latest".to_string(), true),
         MatchSemver::Semver((version, _)) => {
-            let url = ctry!(
-                req,
-                Url::parse(&format!(
-                    "{}/crate/{}/{}",
-                    redirect_base(req),
-                    name,
-                    version
-                )),
-            );
-
-            return Ok(super::cached_redirect(url, CachePolicy::ForeverInCdn));
+            return Ok(super::axum_cached_redirect(
+                &format!("/crate/{}/{}", &params.name, version),
+                CachePolicy::ForeverInCdn,
+            )
+            .into_response());
         }
     };
 
-    let updater = extension!(req, RepositoryStatsUpdater);
-    let details = cexpect!(
-        req,
-        ctry!(
-            req,
-            CrateDetails::new(
-                &mut *conn,
-                name,
-                &version,
-                &version_or_latest,
-                Some(updater)
-            )
-        )
-    );
-
-    let mut res = CrateDetailsPage { details }.into_response(req)?;
-    res.extensions.insert::<CachePolicy>(if is_latest_url {
-        CachePolicy::ForeverInCdn
-    } else {
-        CachePolicy::ForeverInCdnAndStaleInBrowser
-    });
-    Ok(res)
+    let details = spawn_blocking(move || {
+        let mut conn = pool.get()?;
+        Ok(CrateDetails::new(
+            &mut *conn,
+            &params.name,
+            &version,
+            &version_or_latest,
+            Some(&repository_stats_updater),
+        )?
+        .ok_or_else(|| {
+            report_error(&anyhow!("could not find crate details"));
+            AxumNope::VersionNotFound
+        })?)
+    })
+    .await?;
+
+    let mut res = CrateDetailsPage { details }.into_response();
```

Comment:
Rendering the template for this page can involve syntax highlighting code blocks in the readme, I wonder if that takes long enough it should be `spawn_blocking`ed (I'm not entirely sure how, maybe it'd require tagging the `DelayedTemplateRender` as needing the indirection).

Reply:
Definitely a valid concern, will get more pressing with the source pages too. 

Do you mean as an option to the `DelayedTemplateRender`, so in the render-template middleware we are using `spawn_blocking` for the template rendering?